actors/market/tests/market_actor_test.rs

```diff
@@ -1498,6 +1498,186 @@ fn fail_when_deal_update_epoch_is_in_the_future() {
+    let worker_addr = Address::new_id(WORKER_ID);
+    let control_addr = Address::new_id(CONTROL_ID);
+
+    let mut rt = setup();
+
+    let deal_id1 = publish_and_activate_deal(
+        &mut rt,
+        client_addr,
+        provider_addr,
+        owner_addr,
+        worker_addr,
+        control_addr,
+        start_epoch,
+        end_epoch,
+        0,
+        sector_expiry,
+    );
+    let d1 = get_deal_proposal(&mut rt, deal_id1);
+
+    let deal_id2 = publish_and_activate_deal(
+        &mut rt,
+        client_addr,
+        provider_addr,
+        owner_addr,
+        worker_addr,
+        control_addr,
+        start_epoch + 1,
+        end_epoch + 1,
+        0,
+        sector_expiry,
+    );
+
+    // slash deal1
+    let slash_epoch = process_epoch(start_epoch, deal_id2) + ChainEpoch::from(100);
+    rt.set_epoch(slash_epoch);
+    terminate_deals(&mut rt, provider_addr, &[deal_id1]);
+
+    // cron tick will slash deal1 and make payment for deal2
+    let current = slash_epoch + ChainEpoch::from(1);
+    rt.set_epoch(current);
+    rt.expect_send(
+        *BURNT_FUNDS_ACTOR_ADDR,
+        METHOD_SEND,
+        RawBytes::default(),
+        d1.provider_collateral.clone(),
+        RawBytes::default(),
+        ExitCode::OK,
+    );
+    cron_tick(&mut rt);
+
+    assert_deal_deleted(&mut rt, deal_id1, d1);
+    let s2 = get_deal_state(&mut rt, deal_id2);
+    assert_eq!(current, s2.last_updated_epoch);
+    // TODO: actor.checkState(rt)
+}
+
+#[test]
+fn cannot_publish_the_same_deal_twice_before_a_cron_tick() {
+    let start_epoch = ChainEpoch::from(50);
+    let end_epoch = start_epoch + 200 * EPOCHS_IN_DAY;
+
+    let client_addr = Address::new_id(CLIENT_ID);
+    let provider_addr = Address::new_id(PROVIDER_ID);
+    let owner_addr = Address::new_id(OWNER_ID);
+    let worker_addr = Address::new_id(WORKER_ID);
+    let control_addr = Address::new_id(CONTROL_ID);
+
+    // Publish a deal
+    let mut rt = setup();
+    generate_and_publish_deal(
+        &mut rt,
+        client_addr,
+        provider_addr,
+        owner_addr,
+        worker_addr,
+        control_addr,
+        start_epoch,
+        end_epoch,
+    );
+
+    // now try to publish it again and it should fail because it will still be in pending state
+    let d2 = generate_deal_and_add_funds(
+        &mut rt,
+        client_addr,
+        provider_addr,
+        owner_addr,
+        worker_addr,
+        start_epoch,
+        end_epoch,
+    );
+    let buf = RawBytes::serialize(d2.clone()).expect("failed to marshal deal proposal");
+    let sig = Signature::new_bls("does not matter".as_bytes().to_vec());
+    let params = PublishStorageDealsParams {
+        deals: vec![ClientDealProposal { proposal: d2.clone(), client_signature: sig.clone() }],
+    };
+    rt.expect_validate_caller_type(vec![*ACCOUNT_ACTOR_CODE_ID, *MULTISIG_ACTOR_CODE_ID]);
+    let return_value = ext::miner::GetControlAddressesReturnParams {
```

Comment:
nit: would reduce noise and might be useful to have a subfunction for expecting control address send similar to expect_query_network_info